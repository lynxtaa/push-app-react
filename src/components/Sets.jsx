import React from 'react'
import PropTypes from 'prop-types'
import { Redirect } from 'react-router'

import { getSets } from '../store/scheduleSelectors'

import NavPills from './NavPills'
import PushList from './PushList'

const linkFromDay = week => day => ({ link: `/${week}/${day}`, text: `Day ${day}` })

const Sets = ({ match }) => {
	const { day, week } = match.params

	return day ? (
		<div>
			<NavPills links={[1, 2, 3].map(linkFromDay(week))} />
			<PushList sets={getSets(week, day)} />
		</div>
	) : <Redirect to={`/${week}/1`} />
}

Sets.propTypes = {
	match: PropTypes.object.isRequired,
}

export default Sets
