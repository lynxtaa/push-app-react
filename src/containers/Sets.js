import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { Redirect } from 'react-router'
import { set } from 'idb-keyval'

import NavPills from '@components/NavPills'
import PushListContainer from './PushListContainer'
import useFetchedData from '../hooks/useFetchedData'

const Sets = ({ match }) => {
	const [schedule, error] = useFetchedData('/api/schedule')

	useEffect(() => {
		set('route', match.params)
	}, [match.params])

	const { day, week } = match.params

	if (error) {
		return <h2 className="text-danger">{error.message}</h2>
	}

	if (!schedule) {
		return <h2>Loading...</h2>
	}

	if (!day) {
		return <Redirect to={`/${week}/1`} />
	}

	const sets = schedule
		.find(({ weekNum }) => weekNum == week)
		.days[day - 1].map((set, i) => ({ id: `${week}-${day}-${i}`, set }))

	const links = [1, 2, 3].map(day => ({
		link: `/${week}/${day}`,
		text: `Day ${day}`,
	}))

	return (
		<div>
			<NavPills links={links} />
			<PushListContainer sets={sets} />
		</div>
	)
}

Sets.propTypes = {
	match: PropTypes.object.isRequired,
}

export default Sets
