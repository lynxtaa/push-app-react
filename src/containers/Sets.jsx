import React from 'react'
import PropTypes from 'prop-types'
import { Redirect } from 'react-router'

import { saveState } from '../store/localStorage'

import NavPills from '@components/NavPills'
import PushListContainer from './PushListContainer'

class Sets extends React.Component {
	constructor(props) {
		super(props)
		this.state = { schedule: null }
	}

	componentDidMount() {
		fetch('/api/schedule')
			.then(res => res.json())
			.then(schedule => {
				this.setState({ schedule })
			})
	}

	componentWillUpdate({ match }) {
		saveState(match.params)
	}

	get links() {
		const { week } = this.props.match.params

		return [1, 2, 3].map(day => ({
			link: `/${week}/${day}`,
			text: `Day ${day}`,
		}))
	}

	render() {
		const { day, week } = this.props.match.params
		const { schedule } = this.state

		if (!schedule) {
			return <h2>Loading...</h2>
		}

		if (!day) {
			return <Redirect to={`/${week}/1`} />
		}

		const sets = schedule
			.find(({ weekNum }) => weekNum == week).days[day - 1]
			.map((set, i) => ({ id: `${week}-${day}-${i}`, set }))

		return (
			<div>
				<NavPills links={this.links} />
				<PushListContainer sets={sets} />
			</div>
		)
	}
}

Sets.propTypes = {
	match: PropTypes.object.isRequired,
}

export default Sets
