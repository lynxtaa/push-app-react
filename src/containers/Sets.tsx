import React, { useEffect } from 'react'
import { Redirect, useParams } from 'react-router-dom'
import { set } from 'idb-keyval'

import NavPills from 'components/NavPills'
import PushListContainer from './PushListContainer'
import useFetchedData from '../hooks/useFetchedData'

type ScheduleItem = {
	weekNum: number
	days: number[][]
}

function Sets() {
	const [schedule, error] = useFetchedData<ScheduleItem[]>(
		`${process.env.PUBLIC_URL}/schedule.json`,
	)
	const matchParams = useParams<{ day?: string; week: string }>()

	const { day, week } = matchParams

	useEffect(() => {
		set('route', { day, week })
	}, [day, week])

	if (error) {
		return <h2 className="text-danger">{error.message}</h2>
	}

	if (!schedule) {
		return <h2>Loading...</h2>
	}

	if (!day) {
		return <Redirect to={`/${week}/1`} />
	}

	const weekSchedule = schedule.find(({ weekNum }) => weekNum === Number(week))

	if (!weekSchedule) {
		throw new Error(`Week ${week} not found in schedule`)
	}

	const sets = weekSchedule.days[Number(day) - 1].map((set, i) => ({
		id: `${week}-${day}-${i}`,
		set,
	}))

	const links = [1, 2, 3].map(day => ({
		link: `/${week}/${day}`,
		text: `Day ${day}`,
	}))

	return (
		<div>
			<NavPills links={links} />
			<PushListContainer sets={sets} />
		</div>
	)
}

export default Sets
