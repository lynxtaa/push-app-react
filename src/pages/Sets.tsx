import { useEffect } from 'react'
import { Redirect, useParams } from 'react-router-dom'
import { set } from 'idb-keyval'

import NavPills from 'components/NavPills'
import PushList from 'components/PushList'
import useFetchedData from '../hooks/useFetchedData'
import { Text, Box } from '@chakra-ui/core'
import Alert from 'components/Alert'

type ScheduleItem = {
	weekNum: number
	days: number[][]
}

export default function Sets() {
	const [schedule, error] = useFetchedData<ScheduleItem[]>(
		`${process.env.PUBLIC_URL}/schedule.json`,
	)

	const { day, week } = useParams<{ day?: string; week: string }>()

	useEffect(() => {
		set('route', { day, week })
	}, [day, week])

	if (error) {
		return (
			<Alert status="error" onClose={() => window.location.reload()}>
				{error.message}
			</Alert>
		)
	}

	if (!schedule) {
		return <Text fontSize="lg">Loading...</Text>
	}

	if (!day) {
		return <Redirect to={`/${week}/1`} />
	}

	const weekSchedule = schedule.find(({ weekNum }) => weekNum === Number(week))

	if (!weekSchedule) {
		throw new Error(`Week ${week} not found in schedule`)
	}

	const sets = weekSchedule.days[Number(day) - 1].map((set, i) => ({
		id: `${week}-${day}-${i}`,
		set,
	}))

	const links = [1, 2, 3].map((day) => ({
		link: `/${week}/${day}`,
		text: `Day ${day}`,
	}))

	return (
		<Box>
			<NavPills links={links} />
			<PushList sets={sets} />
		</Box>
	)
}
