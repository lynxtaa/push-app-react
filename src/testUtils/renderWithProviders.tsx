import { render, RenderOptions, RenderResult } from '@testing-library/react'
import { ChakraProvider } from '@chakra-ui/react'
import { MemoryRouter } from 'react-router-dom'

import theme from '../theme'

export default function renderWithProviders(
	ui: React.ReactElement,
	{
		initialEntries,
		initialIndex,
		...rest
	}: Omit<RenderOptions, 'queries'> & {
		initialEntries?: string[]
		initialIndex?: number
	} = {},
): RenderResult {
	return render(ui, {
		...rest,
		wrapper: ({ children }: { children?: React.ReactNode }) => (
			<ChakraProvider theme={theme}>
				<MemoryRouter initialEntries={initialEntries} initialIndex={initialIndex}>
					{children}
				</MemoryRouter>
			</ChakraProvider>
		),
	})
}
