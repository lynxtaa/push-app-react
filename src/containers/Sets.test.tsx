import React from 'react'
import { screen } from '@testing-library/react'
import Sets from './Sets'
import { Route } from 'react-router-dom'
import { set } from 'idb-keyval'
import renderWithProviders from 'testUtils/renderWithProviders'

const setMock = set as jest.Mock
const fetchMock = window.fetch as jest.Mock

it('renders sets page', async () => {
	fetchMock.mockResolvedValue({
		ok: true,
		json: () =>
			Promise.resolve([
				{
					weekNum: 1,
					days: [
						[10, 12, 7, 7, 9],
						[10, 12, 8, 8, 12],
						[11, 15, 9, 9, 13],
					],
				},
			]),
	})

	const { container } = renderWithProviders(
		<Route path="/:week(\d)/:day(\d)?">
			<Sets />
		</Route>,
		{ initialEntries: ['/1/1'], initialIndex: 0 },
	)

	await screen.findByText(/loading/i)

	await screen.findByText('12')

	expect(setMock).toHaveBeenCalledWith('route', { week: '1', day: '1' })

	expect(container.firstChild).toMatchSnapshot()
})

it('if fetch resulted in error, shows error', async () => {
	fetchMock.mockRejectedValue(new Error('Fetching Error'))

	renderWithProviders(
		<Route path="/:week(\d)/:day(\d)?">
			<Sets />
		</Route>,
		{ initialEntries: ['/1/1'], initialIndex: 0 },
	)

	await screen.findByText('Fetching Error')
})

it('if no day in match.params, redirects to /<WEEK>/1', async () => {
	fetchMock.mockResolvedValue({
		ok: true,
		json: () =>
			Promise.resolve([
				{
					weekNum: 3,
					days: [
						[10, 12, 7, 7, 9],
						[10, 12, 8, 8, 12],
						[11, 15, 9, 9, 13],
					],
				},
			]),
	})

	renderWithProviders(
		<Route path="/:week(\d)/:day(\d)?">
			<Sets />
		</Route>,
		{ initialEntries: ['/3'], initialIndex: 0 },
	)

	await screen.findByText('9')
})
