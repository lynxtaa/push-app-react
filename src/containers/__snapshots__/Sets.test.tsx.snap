// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if fetch resulted in error, shows error 1`] = `
<h2
  class="text-danger"
>
  Fetching Error
</h2>
`;

exports[`if no day in match.params, redirects to /<WEEK>/1 1`] = `
<div>
  <ul
    class="nav nav-pills mb-3"
  >
    <li
      class="nav-item"
    >
      <a
        aria-current="page"
        class="nav-link active"
        href="/3/1"
      >
        Day 1
      </a>
    </li>
    <li
      class="nav-item"
    >
      <a
        class="nav-link"
        href="/3/2"
      >
        Day 2
      </a>
    </li>
    <li
      class="nav-item"
    >
      <a
        class="nav-link"
        href="/3/3"
      >
        Day 3
      </a>
    </li>
  </ul>
  <ul
    class="pushlist list-unstyled"
  >
    <li
      style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
    >
      <button
        class="btn btn-outline-secondary w-100"
        type="button"
      >
        10
      </button>
    </li>
    <li
      style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
    >
      <button
        class="btn btn-outline-secondary w-100"
        type="button"
      >
        12
      </button>
    </li>
    <li
      style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
    >
      <button
        class="btn btn-outline-secondary w-100"
        type="button"
      >
        7
      </button>
    </li>
    <li
      style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
    >
      <button
        class="btn btn-outline-secondary w-100"
        type="button"
      >
        7
      </button>
    </li>
    <li
      style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
    >
      <button
        class="btn btn-outline-secondary w-100"
        type="button"
      >
        9
      </button>
    </li>
  </ul>
</div>
`;

exports[`renders sets page 1`] = `
<h2>
  Loading...
</h2>
`;

exports[`renders sets page 2`] = `
<div>
  <ul
    class="nav nav-pills mb-3"
  >
    <li
      class="nav-item"
    >
      <a
        aria-current="page"
        class="nav-link active"
        href="/1/1"
      >
        Day 1
      </a>
    </li>
    <li
      class="nav-item"
    >
      <a
        class="nav-link"
        href="/1/2"
      >
        Day 2
      </a>
    </li>
    <li
      class="nav-item"
    >
      <a
        class="nav-link"
        href="/1/3"
      >
        Day 3
      </a>
    </li>
  </ul>
  <ul
    class="pushlist list-unstyled"
  >
    <li
      style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
    >
      <button
        class="btn btn-outline-secondary w-100"
        type="button"
      >
        10
      </button>
    </li>
    <li
      style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
    >
      <button
        class="btn btn-outline-secondary w-100"
        type="button"
      >
        12
      </button>
    </li>
    <li
      style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
    >
      <button
        class="btn btn-outline-secondary w-100"
        type="button"
      >
        7
      </button>
    </li>
    <li
      style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
    >
      <button
        class="btn btn-outline-secondary w-100"
        type="button"
      >
        7
      </button>
    </li>
    <li
      style="opacity: 0; margin-bottom: 0px; height: 0px; visibility: hidden;"
    >
      <button
        class="btn btn-outline-secondary w-100"
        type="button"
      >
        9
      </button>
    </li>
  </ul>
</div>
`;
